lib/rules/sort-comp.js

```diff
@@ -131,84 +131,91 @@ module.exports = {
      * @returns {Array} The matching patterns indexes. Return [Infinity] if there is no match.
      */
     function getRefPropIndexes(method) {
-      let isRegExp;
-      let matching;
-      let i;
-      let j;
-      const indexes = [];
-
-      if (method.getter) {
-        const getterIndex = methodsOrder.indexOf('getters');
-        if (getterIndex >= 0) {
-          indexes.push(getterIndex);
-        }
-      }
-
-      if (method.setter) {
-        const setterIndex = methodsOrder.indexOf('setters');
-        if (setterIndex >= 0) {
-          indexes.push(setterIndex);
-        }
-      }
-
-      if (method.typeAnnotation) {
-        const annotationIndex = methodsOrder.indexOf('type-annotations');
-        if (annotationIndex >= 0) {
-          indexes.push(annotationIndex);
-        }
-      }
+      const methodGroupIndexes = [];
 
-      if (indexes.length === 0) {
-        for (i = 0, j = methodsOrder.length; i < j; i++) {
-          isRegExp = methodsOrder[i].match(regExpRegExp);
-          if (isRegExp) {
-            matching = new RegExp(isRegExp[1], isRegExp[2]).test(method.name);
-          } else {
-            matching = methodsOrder[i] === method.name;
+      methodsOrder.forEach((currentGroup, groupIndex) => {
+        if (currentGroup === 'getters') {
+          if (method.getter) {
+            methodGroupIndexes.push(groupIndex);
           }
-          if (matching) {
-            indexes.push(i);
+        } else if (currentGroup === 'setters') {
+          if (method.setter) {
+            methodGroupIndexes.push(groupIndex);
+          }
+        } else if (currentGroup === 'type-annotations') {
+          if (method.typeAnnotation) {
+            methodGroupIndexes.push(groupIndex);
+          }
+        } else if (currentGroup === 'static-methods') {
+          if (method.static) {
+            methodGroupIndexes.push(groupIndex);
+          }
+        } else if (currentGroup === 'instance-variables') {
+          if (method.instanceVariable) {
+            methodGroupIndexes.push(groupIndex);
+          }
+        } else if (currentGroup === 'instance-methods') {
+          if (method.instanceMethod) {
+            methodGroupIndexes.push(groupIndex);
+          }
+        } else if ([
+          'displayName',
+          'propTypes',
+          'contextTypes',
+          'childContextTypes',
+          'mixins',
+          'statics',
+          'defaultProps',
+          'constructor',
+          'getDefaultProps',
+          'state',
+          'getInitialState',
+          'getChildContext',
+          'getDerivedStateFromProps',
+          'componentWillMount',
+          'UNSAFE_componentWillMount',
```

Comment:
I think the UNSAFE_ methods should only be included in this array when the React version in settings is 16.3+ - same with gDSFP, gSBU, and cDC (for 16.0+)

Reply:
I disagree with this.  In my opinion this makes sense in the config code, but not in the rule implementation itself.

If someone has their config set up so that it includes the `UNSAFE_` methods and they have the `UNSAFE_` methods in their component then eslint should enforce that config no matter which version of React they are using.

Reply:
If someone is using React 15, React 16 lifecycle methods should not be sorted in the lifecycle group - because for them, they aren't lifecycle methods. Similarly, if someone is using React 17, where `componentWillMount` theoretically does not exist, cWM should not be sorted in the lifecycle group - because for them, it's not a lifecycle method.

Reply:
Agreed. This is a concern for the code that loads the default config. 

This code is dealing with the finalised config. It should not undo a config based on assumptions about the environment.  eg the user may have their own `UNSAFE_` method that just happens to match the choice a new version of React implements. One would hope this is unlikely, but it would be bad practice to assume it impossible.  

Assuming we even wanted to try and undo the config, because custom groups are flattened in the config, and someone could overwrite the lifecycle group, we could only undo methods that were not contiguous with other lifecycle methods. Even then we would be assuming that the user used the lifecycle group when they may have hand coded the list.  We could inspect the raw config and deconstruct from there... <pauses> Why are we trying to undo the users config?

Reply:
i guess i'm confused why we need a separate list of lifecycle methods, if this is just "the user's config". should this not fall down to line 200?

Reply:
Yes, line 162 is a functional subset of 200. However 162 is documented behaviour and 200 is not. As stated in the PR description, I've added a test to make sure the functionality isn't lost, but presumed it was better to code to expected behaviours.

Would you prefer that I comment L#200 making it explicit that it's undocumented functionality, or remove 162 and effectively hide (in my opinion) documented behaviour in the code for undocumented behaviour?  Of course the tests would catch if someone removed the code thinking they were removing the undocumented behaviour, but that feels too obfuscated to me.

Reply:
Hmm.

Let's keep this as-is; thanks for your explanation.